From: Daniel van Vugt <daniel.van.vugt@canonical.com>
Date: Fri, 29 Aug 2025 16:09:43 +0800
Subject: compositor: Don't assume there's an active ClutterFrame during X11
 paint

The paint_context created in `create_framebuffer_from_window_actor` is
too short lived to have a `ClutterFrame` associated with it. That only
exists during a view paint. So get the same frame counter from the
intended source (the frame clock) instead.

Bug: https://gitlab.gnome.org/GNOME/mutter/-/issues/4254
Bug-Ubuntu: https://bugs.launchpad.net/bugs/2121195
Origin: https://gitlab.gnome.org/GNOME/mutter/-/merge_requests/4612
Fixes: b5ca2a46d52 ("compositor,x11: Use per-view frame counter ...")
---
 src/compositor/meta-window-actor-x11.c | 18 +++++++++---------
 1 file changed, 9 insertions(+), 9 deletions(-)

diff --git a/src/compositor/meta-window-actor-x11.c b/src/compositor/meta-window-actor-x11.c
index bf1409d..b48248e 100644
--- a/src/compositor/meta-window-actor-x11.c
+++ b/src/compositor/meta-window-actor-x11.c
@@ -186,7 +186,7 @@ queue_send_frame_messages_timeout (MetaWindowActorX11 *actor_x11)
 
 static void
 assign_frame_counter_to_frames (MetaWindowActorX11 *actor_x11,
-                                ClutterFrame       *clutter_frame)
+                                int64_t             frame_count)
 {
   MetaWindow *window =
     meta_window_actor_get_meta_window (META_WINDOW_ACTOR (actor_x11));
@@ -200,14 +200,12 @@ assign_frame_counter_to_frames (MetaWindowActorX11 *actor_x11,
     return;
 
   sync_counter = meta_window_x11_get_sync_counter (window);
-  meta_sync_counter_assign_counter_to_frames (sync_counter,
-                                              clutter_frame->frame_count);
+  meta_sync_counter_assign_counter_to_frames (sync_counter, frame_count);
   frame = meta_window_x11_get_frame (window);
   if (frame)
     {
       sync_counter = meta_frame_get_sync_counter (frame);
-      meta_sync_counter_assign_counter_to_frames (sync_counter,
-                                                  clutter_frame->frame_count);
+      meta_sync_counter_assign_counter_to_frames (sync_counter, frame_count);
     }
 }
 
@@ -1218,7 +1216,7 @@ meta_window_actor_x11_before_paint (MetaWindowActor  *actor,
 
   handle_updates (actor_x11);
 
-  assign_frame_counter_to_frames (actor_x11, frame);
+  assign_frame_counter_to_frames (actor_x11, frame->frame_count);
 }
 
 static void
@@ -1236,11 +1234,13 @@ meta_window_actor_x11_paint (ClutterActor        *actor,
   * and send the completion events normally */
   if (actor_x11->send_frame_messages_timer != 0)
     {
-      ClutterFrame *frame;
+      ClutterFrameClock *frame_clock;
+      int64_t last_count;
 
       remove_frame_messages_timer (actor_x11);
-      frame = clutter_paint_context_get_frame (paint_context);
-      assign_frame_counter_to_frames (actor_x11, frame);
+      frame_clock = clutter_actor_pick_frame_clock (actor, NULL);
+      last_count = clutter_frame_clock_get_frame_count (frame_clock);
+      assign_frame_counter_to_frames (actor_x11, last_count);
     }
 
   window = meta_window_actor_get_meta_window (META_WINDOW_ACTOR (actor_x11));
